### DISCUSSION TITLE
simulate restrained inflatable with Sofa

### DISCUSSION THREAD

**Discussion by duchaoyu** [sofa-framework/sofa#4276] [Write / Run a simulation] [answered]
Hi, I'm new to Sofa and from the field of engineering and architecture. I want to use Sofa to simulate a more realistic restrained inflatable. I wanna achieve something like this: a fabric lifted up by an elastic balloon (image from Blender). 
<img width="662" alt="image" src="https://github.com/sofa-framework/sofa/assets/43376817/10a0f8dc-f014-4028-aed7-5998d67f6c83">

Here is the first outcome I achieved from Sofa. 
![image](https://github.com/sofa-framework/sofa/assets/43376817/caafbaef-9abf-47ee-b096-d01979839d6d)
I don't know how to anchor the fabric on the ground. Thus, it's currently floating in the air. But I assume if the tensioned fabric is at the ground, the balloon will be pressed down and becomes oval. 

Here are my questions: 
1. in the first a few frames, my mesh will explode. But later the mesh will be stable again. I have tried to make the time step smaller, but the problem remains. What can be reason of this explosion? 
<img width="650" alt="image" src="https://github.com/sofa-framework/sofa/assets/43376817/48045fcc-2c19-4e60-bb6c-7402e0ad81d4">
2. How can I always keep the inflatable beneath the fabric? Now I'm using a sphere as an input mesh. Should I firstly deflate my balloon, export it, and use it as the input mesh?
3. How can I modify the tension in some continuous edges? (to represent material property of a cable)

Here are the code and the obj files. 
[restrained_pneu.zip](https://github.com/sofa-framework/sofa/files/13228746/restrained_pneu.zip)

**Comment by bakpaul** [comment, marked as answer]
Here in your scene you are using penalty-based collision response. This means that the interaction is modeled by springs added between the colliding objects. As you certainly know, a spring has got two parameters : a rest length (here it is set by the contact distance) and the stiffness. This stiffness is not set in your scene and thus a default value is provided. I suggest therefore to modify your lines  169-172 in the following way : 

``` py 
    f_trianglesMesh.addObject('TriangleCollisionModel', name="Models", contactStiffness="0.1" )
    # trianglesMesh.addObject('SurfacePressureForceField', pressure=200.0, pulseMode=True, pressureSpeed=20.0)
    f_trianglesMesh.addObject('LineCollisionModel', contactStiffness="0.1")
    f_trianglesMesh.addObject('PointCollisionModel', contactStiffness="0.1")
```

Now you don't have explosion anymore, but if you look closer you can still see some rebounds. In my knowledge, you'll not be able to do really better with penalty on such lightweight object. You have two solutions I can think of : 
1. Make your cloth heavier (if you don't care about the dynamics and only the static, this may not be a problem for you.)
2. Use Langrangian-based constraints (see [this scene for a simple example](https://github.com/sofa-framework/sofa/blob/master/examples/Demos/fallingBeamLagrangianCollision.scn))

**Comment by duchaoyu** [comment]
Dear Paul, thank you so much for your answer, and I apologize for my delayed response. Amazing! Thanks to your guidance, the simulation is now functioning correctly!

I have looked into the Langrangian-based constraints. If I understand correctly, the Langrangian-based constraints can be more precise than the penalty-based collision response. I have attempted to adapt `3.2_cloth_penalty.py` into a new file named `3.2_cloth_lagrangian.py`. This modification was inspired by the examples from the SOFA library, because I don't think I fully understand the difference... However, I've encountered an issue where the cloth in the modified simulation remains static and shows no movement. Would it be possible for you to take a moment to have a look and suggest what might be going wrong?

Besides, I tried to simulate the inflation of a deflated balloon with a cloth (or more accurately, a thin shell) placed on top (`3.3_inflate_cloth.py`). The cloth/shell is envisioned to be heavy, for example being coated with a thin layer of wet concrete, yet retaining flexibility. I think I have almost achieved what I had in mind . However, penetration still happens in some regions (see image below). I have tried modifying the time step, and the collision distance and contact stiffness. However, the penetration remains. I wonder whether using Langrangian-based constraints will solve the problems? Or would it be more effective to model the heavy cloth using tetrahedron finite elements instead of mesh elements?

[sofa_restrained_inflatable.zip](https://github.com/sofa-framework/sofa/files/13498820/sofa_restrained_inflatable.zip)
![image](https://github.com/sofa-framework/sofa/assets/43376817/83320a51-a88f-4356-99cc-644feab5e60a)

Thank you very much for your reply!

**Reply by bakpaul** [reply]
Interpenetration is difficult to avoid using penalty, even more when using a 2D objects, because of the lack of thickness. Because of that, even a slight penetration will result in the point being on the other side of you object and thus the penalty pushing on the wrong direction. 

The Lagrangian formulation would suffer of the same problem when the constraint solver won't converge, which might be the case here because your fabric is so lightweight, but when it converges you have a garantee of no interpenetration. 

One first step before the lagrangian multiplier would be to manually change the surface normal of your cloth used for the collision, so that triangles from both sides are pointing to the ground, so that penalty are always rightly oriented. I don't have a simple way to do it, I've never done it, but it might be easy. 

The second test would be indeed to add thinckness to your cloth by meshing it. You could use a SpaseGrid to mesh it easily. 

Finally, I didn't have the time to test your scene with langrangian multiplirs, but looking at it, I can already see one mistake : you don't have a direct solver in your mechanical nodes. You only have `CGDirectSolver`. You need a direct solver (such as `SparseLDLSolver`) to use the `LinearSolverConstraintCorrection`, and you should link to it (`linearSolver=@yoursolvername`)
